test/jdk/javax/swing/JMenu/PopupReferenceMemoryLeak.java

```diff
@@ -0,0 +1,179 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 4907798
+ * @key headful
+ * @requires (os.family == "windows")
+ * @summary Check for memory leak in menu subsystem
+ * @run main/othervm PopupReferenceMemoryLeak
+ */
+
+import javax.swing.AbstractAction;
+import javax.swing.JFrame;
+import javax.swing.JMenu;
+import javax.swing.JMenuBar;
+import javax.swing.JMenuItem;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+import javax.swing.WindowConstants;
+import java.awt.BorderLayout;
+import java.awt.Robot;
+import java.awt.event.ActionEvent;
+import java.awt.event.InputEvent;
+import java.awt.event.KeyEvent;
+import java.lang.ref.WeakReference;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+
+public class PopupReferenceMemoryLeak {
+    static volatile WeakReference referenceToFrame1;
+    static JFrame frame1, frame2;
+
+    public static void main(String[] args) throws InvocationTargetException, InterruptedException {
+        try {
+            // Set system look and feel
+            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
+        } catch (Exception ignore) {
+            throw new RuntimeException("Can not install system look and feel");
+        }
+
+        PopupReferenceMemoryLeak newTest = new PopupReferenceMemoryLeak();
+        SwingUtilities.invokeAndWait(newTest::createUI);
+        try {
+            Robot robot = new Robot();
+            robot.setAutoDelay(200);
+            robot.waitForIdle();
+            Thread.sleep(2000);
+            robot.mouseMove(200, 200);
+            robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+            robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+            robot.keyPress(KeyEvent.VK_ALT);
+            robot.keyPress(KeyEvent.VK_F);
+            robot.keyRelease(KeyEvent.VK_F);
+            robot.keyRelease(KeyEvent.VK_ALT);
+            robot.keyPress(KeyEvent.VK_C);
+            robot.keyRelease(KeyEvent.VK_C);
+            robot.waitForIdle();
+            Thread.sleep(2000);
+            robot.mouseMove(600, 200);
+            robot.waitForIdle();
+            Thread.sleep(2000);
+            robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+            robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+
+            // Force GC three times to see if it clears the old frame
+            for(int i=0; i<3; i++) {
+                try {
+                    ArrayList gc = new ArrayList();
+                    while (true) {gc.add(new int[100000]);}
+                } catch (Throwable ignore){}
+                robot.waitForIdle();
+                Thread.sleep(1000);
+                if(referenceToFrame1.get() == null) {
+                    // Frame was released
+                    return;
+                }
+            }
+            robot.waitForIdle();
+            robot.keyPress(KeyEvent.VK_ALT);
```

Comment:
Can we skip the robot interaction and instead show the popup menu ourselves? (not via menu items)

Reply:
No, unfortunately one of the mechanics i'm testing only reproducible when popup is being opened from JMenuBar.